File: frontend/js/views/true_hourly_wage.js
// True hourly wage — net pay after job costs over all job-related hours, via
// /calc/true-hourly-wage. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 1 }) + '%';
const hrs = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const VIEW = 'true-hourly-wage';
let lastReport = null;
let lastBody = null;

export async function renderTrueHourlyWage(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.truewage.h1.title">// TRUE HOURLY WAGE</span></h1>
        <p class="muted small" data-i18n="view.truewage.hint.intro">
            What a job really pays once job-related costs come out of pay and job-related time —
            commute, prep, decompression — is added to the hours. The nominal hourly (salary ÷ paid
            hours) almost always overstates it. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.truewage.h2.inputs">The job</h2>
            <form id="truewage-form" class="inline-form">
                <label><span data-i18n="view.truewage.label.salary">Gross annual salary ($)</span>
                    <input type="number" step="0.01" min="0" name="gross_annual_salary_usd" value="60000" required></label>
                <label><span data-i18n="view.truewage.label.weeks">Weeks worked / year</span>
                    <input type="number" step="1" min="1" name="weeks_worked" value="50" required></label>
                <label><span data-i18n="view.truewage.label.work">Work hours / week</span>
                    <input type="number" step="0.5" min="0" name="weekly_work_hours" value="40" required></label>
                <label><span data-i18n="view.truewage.label.commute">Commute hours / week</span>
                    <input type="number" step="0.5" min="0" name="weekly_commute_hours" value="5"></label>
                <label><span data-i18n="view.truewage.label.extra">Extra job hours / week</span>
                    <input type="number" step="0.5" min="0" name="weekly_extra_hours" value="0"></label>
                <label><span data-i18n="view.truewage.label.taxes">Annual taxes ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_taxes_usd" value="12000"></label>
                <label><span data-i18n="view.truewage.label.expenses">Annual work expenses ($)</span>
                    <input type="number" step="0.01" min="0" name="annual_work_expenses_usd" value="3000"></label>
            </form>
            <div id="truewage-tools" class="ce-toolbar"></div>
            <button type="button" id="truewage-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="truewage-sens" class="ce-sens"></div>
        </div>
        <div id="truewage-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#truewage-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        return {
            gross_annual_salary_usd: Number(fd.get('gross_annual_salary_usd')) || 0,
            weeks_worked: Number(fd.get('weeks_worked')) || 0,
            weekly_work_hours: Number(fd.get('weekly_work_hours')) || 0,
            weekly_commute_hours: Number(fd.get('weekly_commute_hours')) || 0,
            weekly_extra_hours: Number(fd.get('weekly_extra_hours')) || 0,
            annual_taxes_usd: Number(fd.get('annual_taxes_usd')) || 0,
            annual_work_expenses_usd: Number(fd.get('annual_work_expenses_usd')) || 0,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcTrueHourlyWage(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.truewage.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#truewage-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'true-hourly-wage.csv' });
    mount.querySelector('#truewage-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['true_hourly_usd', r.true_hourly_usd],
        ['nominal_hourly_usd', r.nominal_hourly_usd],
        ['erosion_pct', r.erosion_pct],
        ['net_annual_usd', r.net_annual_usd],
        ['total_annual_hours', r.total_annual_hours],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#truewage-result');
    // Line chart: true hourly as weekly commute hours sweep 0 -> 20 (more unpaid time erodes pay).
    const xs = enh.linspace(0, 20, 13);
    const pts = await Promise.all(xs.map(async (c) => {
        const rr = await api.calcTrueHourlyWage({ ...body, weekly_commute_hours: c });
        return { x: c, y: rr ? rr.true_hourly_usd : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'commute h/wk', ylabel: 'true $/hr' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.truewage.h2.result">Your real pay</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.truewage.card.true">True hourly</div>
                    <div class="value pos">${money(r.true_hourly_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.truewage.card.nominal">Nominal hourly</div>
                    <div class="value">${money(r.nominal_hourly_usd)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.truewage.card.erosion">Erosion</div>
                    <div class="value neg">${pct(r.erosion_pct)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.truewage.row.net">Net annual (after job costs)</td><td>${money(r.net_annual_usd)}</td></tr>
                    <tr><td data-i18n="view.truewage.row.nominalhours">Paid hours / year</td><td>${hrs(r.nominal_annual_hours)}</td></tr>
                    <tr><td data-i18n="view.truewage.row.totalhours">Job-related hours / year</td><td>${hrs(r.total_annual_hours)}</td></tr>
                    <tr><td data-i18n="view.truewage.row.nominal">Nominal hourly</td><td>${money(r.nominal_hourly_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.truewage.row.true">True hourly</td><td>${money(r.true_hourly_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#truewage-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: weekly commute hours 0 -> 20; y: annual work expenses 0 -> 2x. Output: true hourly.
    const ex = base.annual_work_expenses_usd || 3000;
    const xVals = enh.linspace(0, 20, 5);
    const yVals = enh.linspace(0, ex * 2, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'weekly_commute_hours', yKey: 'annual_work_expenses_usd', xVals, yVals, compute: (b) => api.calcTrueHourlyWage(b), pick: (r) => (r ? r.true_hourly_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells, fmt: (v) => (v == null ? '—' : '$' + v.toFixed(2)), xfmt: (v) => v.toFixed(0) + 'h', yfmt: (v) => '$' + Math.round(v / 1000) + 'k', xName: t('view.truewage.label.commute') || 'Commute', yName: t('view.truewage.label.expenses') || 'Expenses' });
}

File: crates/traderview-core/src/true_hourly_wage.rs
//! True hourly wage ("Your Money or Your Life") — what a job really pays once
//! job-related costs come out of pay and job-related time (commute, prep,
//! decompression) is added to the hours.
//!
//! ```text
//! net annual  = salary − taxes − work expenses
//! real hours  = (work + commute + extra) per week × weeks
//! true hourly = net annual / real hours
//! ```
//!
//! The nominal hourly (salary ÷ paid hours) almost always overstates it.

use serde::{Deserialize, Serialize};

fn d_weeks() -> f64 {
    50.0
}

#[derive(Debug, Clone, Deserialize)]
pub struct TrueWageInput {
    pub gross_annual_salary_usd: f64,
    #[serde(default = "d_weeks")]
    pub weeks_worked: f64,
    pub weekly_work_hours: f64,
    /// Commute hours per week.
    #[serde(default)]
    pub weekly_commute_hours: f64,
    /// Extra job-related hours per week (prep, decompression, on-call).
    #[serde(default)]
    pub weekly_extra_hours: f64,
    /// Annual taxes on the salary.
    #[serde(default)]
    pub annual_taxes_usd: f64,
    /// Annual job-related expenses (commuting cost, work clothes, meals, etc.).
    #[serde(default)]
    pub annual_work_expenses_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct TrueWageResult {
    /// Salary − taxes − work expenses.
    pub net_annual_usd: f64,
    /// Paid hours per year (work hours × weeks).
    pub nominal_annual_hours: f64,
    /// All job-related hours per year (work + commute + extra × weeks).
    pub total_annual_hours: f64,
    /// Salary / nominal hours.
    pub nominal_hourly_usd: f64,
    /// Net annual / total job-related hours.
    pub true_hourly_usd: f64,
    /// How much the true wage falls short of the nominal, percent.
    pub erosion_pct: f64,
}

pub fn analyze(input: &TrueWageInput) -> TrueWageResult {
    let weeks = input.weeks_worked.max(0.0);
    let net = input.gross_annual_salary_usd - input.annual_taxes_usd - input.annual_work_expenses_usd;

    let nominal_hours = input.weekly_work_hours * weeks;
    let total_hours = (input.weekly_work_hours + input.weekly_commute_hours + input.weekly_extra_hours) * weeks;

    let nominal_hourly = if nominal_hours > 0.0 {
        input.gross_annual_salary_usd / nominal_hours
    } else {
        0.0
    };
    let true_hourly = if total_hours > 0.0 {
        net / total_hours
    } else {
        0.0
    };
    let erosion = if nominal_hourly > 0.0 {
        (1.0 - true_hourly / nominal_hourly) * 100.0
    } else {
        0.0
    };

    TrueWageResult {
        net_annual_usd: net,
        nominal_annual_hours: nominal_hours,
        total_annual_hours: total_hours,
        nominal_hourly_usd: nominal_hourly,
        true_hourly_usd: true_hourly,
        erosion_pct: erosion,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-4
    }

    fn base() -> TrueWageInput {
        TrueWageInput {
            gross_annual_salary_usd: 60_000.0,
            weeks_worked: 50.0,
            weekly_work_hours: 40.0,
            weekly_commute_hours: 5.0,
            weekly_extra_hours: 0.0,
            annual_taxes_usd: 12_000.0,
            annual_work_expenses_usd: 3_000.0,
        }
    }

    #[test]
    fn net_annual() {
        assert!(close(analyze(&base()).net_annual_usd, 45_000.0));
    }

    #[test]
    fn nominal_hourly() {
        // 60,000 / (40 × 50) = 30.
        assert!(close(analyze(&base()).nominal_hourly_usd, 30.0));
    }

    #[test]
    fn true_hourly() {
        // 45,000 / (45 × 50 = 2,250) = 20.
        assert!(close(analyze(&base()).true_hourly_usd, 20.0));
    }

    #[test]
    fn erosion() {
        // 1 − 20/30 = 33.33%.
        assert!(close(analyze(&base()).erosion_pct, 100.0 / 3.0));
    }

    #[test]
    fn hours_breakdown() {
        let r = analyze(&base());
        assert!(close(r.nominal_annual_hours, 2_000.0));
        assert!(close(r.total_annual_hours, 2_250.0));
    }

    #[test]
    fn no_costs_no_commute_equals_nominal() {
        let r = analyze(&TrueWageInput {
            weekly_commute_hours: 0.0,
            annual_taxes_usd: 0.0,
            annual_work_expenses_usd: 0.0,
            ..base()
        });
        assert!(close(r.true_hourly_usd, r.nominal_hourly_usd));
        assert!(close(r.erosion_pct, 0.0));
    }

    #[test]
    fn longer_commute_lowers_true_wage() {
        let short = analyze(&base());
        let long = analyze(&TrueWageInput {
            weekly_commute_hours: 15.0,
            ..base()
        });
        assert!(long.true_hourly_usd < short.true_hourly_usd);
    }

    #[test]
    fn expenses_lower_net_and_wage() {
        let more = analyze(&TrueWageInput {
            annual_work_expenses_usd: 10_000.0,
            ..base()
        });
        assert!(more.true_hourly_usd < analyze(&base()).true_hourly_usd);
    }
}
